rasa/utils/tensorflow/layers.py

```diff
@@ -1085,3 +1140,314 @@ def call(
+
+            x = [[0 1 2],
+                 [3 4 5],
+                 [6 7 8]]
+            candidate_ids = [[0, 1], [0, 0], [2, 0]]
+            gives
+            [
+                [[0 1 2],
+                 [3 4 5]],
+                [[0 1 2],
+                 [0 1 2]],
+                [[6 7 8],
+                 [0 1 2]]
+            ]
+            with one dimension added by tf.expand_dims(candidate_values, axis=1).
+        """
+        tiled_x = tf.tile(
+            tf.expand_dims(layers_utils.batch_flatten(x), 0),
+            (tf.shape(candidate_ids)[0], 1, 1),
+        )
+        candidate_values = tf.gather(tiled_x, candidate_ids, batch_dims=1)
+        candidate_values = tf.expand_dims(candidate_values, axis=1)
+
+        return candidate_values  # (batch_size, 1, num_candidates, tf.shape(x)[-1])
+
+    def _loss_sigmoid(
+        self,
+        sim_pos: tf.Tensor,  # (batch_size, 1, 1)
+        sim_candidates_il: tf.Tensor,  # (batch_size, 1, num_candidates)
+        pos_neg_labels: tf.Tensor,  # (batch_size, num_candidates)
+        mask: Optional[tf.Tensor],
+    ) -> tf.Tensor:  # ()
+        """Computes the sigmoid loss."""
+        # Concatenate the one guaranteed positive example with the candidate examples,
+        # some of which are positives and others are negatives. Which are which
+        # is stored in `pos_neg_labels`.
+        logits = tf.concat([sim_pos, sim_candidates_il], axis=-1, name="logit_concat")
+        logits = tf.squeeze(logits, 1)
+
+        # Create label_ids for sigmoid
+        pos_label_ids = tf.squeeze(tf.ones_like(sim_pos, tf.float32), -1)
+        label_ids = tf.concat(
+            [pos_label_ids, pos_neg_labels], axis=-1, name="gt_concat"
+        )
+
+        # Compute the sigmoid cross-entropy loss. When minimized, the embeddings
+        # for the two classes (positive and negative) are pushed away from each
+        # other in the embedding space, while it is allowed that any input embedding
+        # (featurized tracker) corresponds to more than one label (intent).
+        loss = tf.nn.sigmoid_cross_entropy_with_logits(labels=label_ids, logits=logits)
+
+        if self.scale_loss:
+            # In case of cross entropy log_likelihood = -loss
+            loss *= _scale_loss(-loss)
+
+        if mask is not None:
+            loss *= mask
+
+        if len(loss.shape) == 2:
+            # Average over the sequence
+            if mask is not None:
+                loss = tf.reduce_sum(loss, axis=-1) / tf.reduce_sum(mask, axis=-1)
+            else:
+                loss = tf.reduce_mean(loss, axis=-1)
+
+        # Average the loss over the batch
+        return tf.reduce_mean(loss)
+
+    @staticmethod
+    def _accuracy(
+        sim_pos: tf.Tensor,  # (batch_size, 1, 1)
+        sim_candidates: tf.Tensor,  # (batch_size, 1, num_candidates)
+        pos_neg_labels: tf.Tensor,  # (batch_size, num_candidates)
```

Comment:
This one as well can be made to reuse between the two classes. For `SingleLabelDotProductLoss`, we can pass `sim_pos` as `pos_neg_labels`.